src/simulation/m_lag_bubbles.fpp

```diff
@@ -0,0 +1,1746 @@
+                            if (q_particle%vf(1)%sf(i, j, k) > (1.0d0 - lag_valmaxvoid)) then
+                                rhs_vf(l)%sf(i, j, k) = rhs_vf(l)%sf(i, j, k) + &
+                                                        q_cons_vf(l)%sf(i, j, k)/q_particle%vf(1)%sf(i, j, k)* &
+                                                        q_particle%vf(2)%sf(i, j, k)
+                            end if
+                        end do
+                    end do
+                end do
+            end do
+        end if
+
+        do l = 1, num_dims
+
+            call s_gradient_dir(q_prim_vf(E_idx), q_particle%vf(3), l)
+
+            !$acc parallel loop collapse(3) gang vector default(present)
+            do k = 0, p
+                do j = 0, n
+                    do i = 0, m
+                        if (q_particle%vf(1)%sf(i, j, k) > (1.0d0 - lag_valmaxvoid)) then
+                            rhs_vf(num_fluids + l)%sf(i, j, k) = rhs_vf(num_fluids + l)%sf(i, j, k) - &
+                                                                 (1.0d0 - q_particle%vf(1)%sf(i, j, k))/ &
+                                                                 q_particle%vf(1)%sf(i, j, k)* &
+                                                                 q_particle%vf(3)%sf(i, j, k)
+                        end if
+                    end do
+                end do
+            end do
+
+            !source in energy
+            !$acc parallel loop collapse(3) gang vector default(present)
+            do k = idwbuff(3)%beg, idwbuff(3)%end
+                do j = idwbuff(2)%beg, idwbuff(2)%end
+                    do i = idwbuff(1)%beg, idwbuff(1)%end
+                        q_particle%vf(3)%sf(i, j, k) = q_prim_vf(E_idx)%sf(i, j, k)*q_prim_vf(num_fluids + l)%sf(i, j, k)
+                    end do
+                end do
+            end do
+
+            call s_gradient_dir(q_particle%vf(3), q_particle%vf(4), l)
+
+            !$acc parallel loop collapse(3) gang vector default(present)
+            do k = 0, p
+                do j = 0, n
+                    do i = 0, m
+                        if (q_particle%vf(1)%sf(i, j, k) > (1.0d0 - lag_valmaxvoid)) then
+                            rhs_vf(E_idx)%sf(i, j, k) = rhs_vf(E_idx)%sf(i, j, k) - &
+                                                        q_particle%vf(4)%sf(i, j, k)*(1.0d0 - q_particle%vf(1)%sf(i, j, k))/ &
+                                                        q_particle%vf(1)%sf(i, j, k)
+                        end if
+                    end do
+                end do
+            end do
+        end do
+
+    end subroutine s_add_sources
+
+    !>  This subroutine computes the Keller-Miksis equation
+        !! @param nparticles Particle identifier
+        !! @param step Current time-stage in the 4th/5th order Runge—Kutta–Cash–Karp time-stepping algorithm
+        !! @param q_prim_vf Primitive variables
+    subroutine s_compute_KM(nparticles, step, q_prim_vf)
```

Comment:
I would have to imagine that some of this code duplicates what is in `m_bubbles` for computing Keller--Miksis. is there a way to consolidate the bubble models for the radial dynamics down to one routine?

Reply:
Yes, there was code duplication here. I was able to utilize what was already available for the EE model. Fixed